SCClassLibrary/JITLib/Patterns/Pdef.sc

```diff
@@ -538,10 +538,12 @@ EventPatternProxy : TaskProxy {
 					]).asStream
 				}
 			}{
+				var fadeOutCleanup = EventStreamCleanup.new.functions_(cleanup.functions.copy);
```

Comment:
it would be nicer if you added a `copy` method to `EventStreamCleanup`:
```supercollider
copy {
			^this.class.new.functions_(this.functions.copy)
}
```

Reply:
Yes that would be nicer, but I have to audit the rest of the classlib code that doesn't somehow use and depend on the default `copy` (inherited from Object) in EventStreamCleanup, i.e. the one that shared the functions array instead of copying it too. The answer is *probably* "no there's no code that does that"... but there's lots of code that uses cleanups...

Reply:
I see what you mean. In the end, you add a workaround for the current `copy` method in `EventStreamCleanup`, which is obviously broken. If you would copy a cleanup, you would not want a shallow copy, would you? 

Btw. I've searched the class library, there is no use of copy there on cleanups.

Reply:
Yes, I did that search too in classlib and also in Daniel's miSC lib, which has somewhat more complicated classes (PmonoPar, PL_Pproxy* etc.) And it seems A-OK to make this change (add that copy with slightly deeper semantics, i.e. make a `functions.copy`.)  I think @jamshark70  will let us know if this were to break any of his libs.

But

>  If you would copy a cleanup, you would not want a shallow copy, would you?

It depends what you mean by the word "cleanup" here. An EventStreamCleanup object is one thing, which nobody seemed to `copy` (that is explicitly) insofar anyway. Explicitly copying a CleanupThunk (the new proposed implementation using "upgraded" Thunks [so they don't ignore arguments on `value`] replacing "naked" function in the `functions` Set) with `copy`  would give two different "cleanups" (CleanupThunks) each with its own "didEvaluate" flag. It's only the fact that CleanupThunks are passed inside an array in the addToCleanup "message" in Events, and nobody does a deep copy on those arrays that makes the proposed fix for the double-cleanup [possible](https://github.com/supercollider/supercollider/issues/4806#issuecomment-632997688), while persevering the current interface to EventStreamCleanup, which lots of classlib code depends on. Being able to *explicitly* `copy` CleanupThunks with "separate didEvaluate" turned out to be potentially beneficial in preserving the PmonoArtic `stop` behavior, although that is not really a "killer application" since that behavior of PmonoArtic is rather undesirable.

The new `EventStreamCleanup.copy` that does a *shallow* copy of its `functions` (Set) seems fine. It wouldn't be at all fine if EventStreamCleanup.copy did a `deepCopy` (even one level-deep) in the `functions` as that would call (or be equivalent with calling) CleanupThunk.copy on every stored item in `functions`.

Reply:
Soo, all those ramblings aside, it turns out that this bugfix works just with

```
var fadeOutCleanup = cleanup.copy; // works even with Object.copy-inherited copy in EventStreamCleanup
cleanup.clear; // change need be seen by caller function, i.e. embedInStream
```
 as James predicted it would, because of the clear on the next line means that previously copied `cleanup` into `fadeOutCleanup` has a different `functions` object.

```
c = EventStreamCleanup.new.addFunction((), {"hi".postln})
d = c.copy

// just checking I've disabled the overriden one
c.class.findRespondingMethodFor(\copy) // -> Object:copy

d.functions === c.functions // true (because of Object.copy!)
c.clear
d.functions === c.functions // false now even with Object.copy
```

So I'm just going to commit this here. If the slightly deeper copy is in ESC.copy is desirable, it can be committed separately as a "breaking change".

Reply:
OK, good!